src/blockchain_db/locked_txn.h

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2014-2019, The Monero Project
+//
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Parts of this file are originally copyright (c) 2012-2013 The Cryptonote developers
+
+#pragma once
+
+namespace cryptonote
+{
+    // This class is meant to create a batch when none currently exists.
+    // If a batch exists, it can't be from another thread, since we can
+    // only be called with the txpool lock taken, and it is held during
+    // the whole prepare/handle/cleanup incoming block sequence.
```

Comment:
How about adding:

It should always be called with the txpool and blockchain locks taken, in that order.

(which might be changed later to allow for rw locks, but for now it is the case)

Reply:
I'm using the following in another patch in txpool.cpp:

```
class RPCLockedTXN
{
public:
  RPCLockedTXN(cryptonote::core &core):
    core(core)
  {
    core.lock_txpool();
    core.get_blockchain_storage().lock();
    child.reset(new cryptonote::LockedTXN(core.get_blockchain_storage().get_db()));
  }
  virtual ~RPCLockedTXN()
  {
    child.reset(nullptr);
    core.get_blockchain_storage().unlock();
    core.unlock_txpool();
  }

  void commit()
  {
    child->commit();
  }

private:
  cryptonote::core &core;
  std::unique_ptr<cryptonote::LockedTXN> child;
};

```

Reply:
I'm not sure if I would be comfortable with access to the database being controlled by the locks on the mempool and the blockchain. It's laying a lot of burden on the programmer to know exactly what class has what responsibility, and it's stretching even further the issue that the code doesn't tell what data a mutex protects.

Having looked a little at the BlockchainLMDB implementation today, I'm somewhat more in favour of making BlockchainLMDB::batch_start(), in addition to all its other methods that are called without a running batch, thread-safe (and block if another thread already has a transaction/batch open). Then LockedTXN doesn't require changes beyond the PR as it was, I believe, and the concurrency issues are tackled on a more basic level.

Reply:
I'm not saying it should be, just that it currently is. You're free to change these constraints of course.

Reply:
Probably the m_batch_active flag should have been made thread-local, since the point of batch_start is to allow a thread to join an existing batch if it's already in progress. After that, merely attempting to start a new write txn would already be a blocking operation inside LMDB, which takes care of the other cases.

Reply:
That sounds like a good idea, indeed. How about I make a PR that does that?

Reply:
Turns out that was slightly over-eager; of course it doesn't make sense to have thread_local class members. m_batch_active could become static, but that would only allow one instance of BlockchainDB.